### ISSUE TITLE
Bug report: make under jparse/ not working right if at all wrt the parser / scanner files

### ISSUE THREAD

**Issue by xexyl** [ioccc-src/mkiocccentry#512]
## Describe the bug

I'm working on adding a prefix to the parser / scanner and I'm running into a problem. I might be wrong but I noticed something funny and I wonder if it's part of the problem.

Look at this:

```sh
$ touch jparse.[yl]
$ make clobber parser-o

jparse: make legacy_clean starting


test_jparse: make legacy_clean starting


test_jparse: make legacy_clean ending
jparse: nothing to do

jparse: make legacy_clean ending

jparse: make legacy_clobber starting


test_jparse: make legacy_clean starting


test_jparse: make legacy_clean ending

test_jparse: make legacy_clobber starting


test_jparse: make legacy_clobber ending
jparse: nothing to do

jparse: make legacy_clobber ending

jparse: make clean starting

rm -f jparse.o jparse.tab.o json_parse.o json_sem.o json_util.o util.o verge.o jsemtblgen.o jstrdecode.o jstrencode.o jparse_main.o jparse.tab.c jparse.tab.h jparse.lex.h

jparse: make clean ending

jparse: make clobber starting

make --no-print-directory -C test_jparse clobber

test_jparse: make legacy_clean starting


test_jparse: make legacy_clean ending

test_jparse: make legacy_clobber starting


test_jparse: make legacy_clobber ending

test_jparse: make clean starting

rm -f jnum_chk.o jnum_test.o jnum_gen.o  

test_jparse: make clean ending

test_jparse: make clobber starting

rm -f jnum_chk jnum_gen
rm -f jparse_test.log chkentry_test.log txzchk_test.log
rm -f tags .local.dir.tags

test_jparse: make clobber ending
rm -f jparse.a jparse verge jsemtblgen jstrdecode jstrencode 
rm -f jparse.output lex.yy.c jparse.c
rm -f jsemcgen.out.*
rm -f tags .local.dir.tags

jparse: make clobber ending
make parser RUN_O_FLAG='-o'
rm -f jparse.tab.c jparse.tab.h
make jparse.tab.c jparse.tab.h
clang -std=gnu11 -O3 -g3 -pedantic -Wall -Wextra   verge.c -c
clang -std=gnu11 -O3 -g3 -pedantic -Wall -Wextra   util.c -c
clang -std=gnu11 -O3 -g3 -pedantic -Wall -Wextra   verge.o util.o ../dbg/dbg.a ../dyn_array/dyn_array.a -o verge
make[2]: `jparse.tab.h' is up to date.
make jparse.tab.o
cat ./jparse.c >> .sorry.run_flex.sh.XXXXXXXXXX.c
cat ./jparse.lex.h >> .sorry.run_flex.sh.XXXXXXXXXX.c
clang -std=gnu11 -O3 -g3 -pedantic -Wall -Wextra   jparse.tab.c -c
rm -f jparse.c jparse.lex.h
make jparse.c jparse.lex.h
cat ./jparse.c >> .sorry.run_flex.sh.XXXXXXXXXX.c
cat ./jparse.lex.h >> .sorry.run_flex.sh.XXXXXXXXXX.c
make[2]: `jparse.lex.h' is up to date.


```

.. but why? Notice that header files are supposedly up to date even though I modified the source files that should force it to be updated. As well it doesn't appear to run `run_flex.sh` or `run_bison.sh` probably because it's up to date (supposedly).

That was from the `jparse/` directory but the same problem from the root directory.

Strangely the ref files are not being updated either probably because the scripts are not running. Thus if I copy the ref file to the target file and run `make` again they will differ with the ref file being the old way. But this is a problem because the ref is supposed to be a backup file.

## To Reproduce

Steps to reproduce the behaviour (adjust steps as needed):

0. Update `jparse.[ly]` or only one of the files.
1. Run `make parser` from either the root directory or `jparse/` itself. `make parser-o` does not work either. As I recall we didn't even use to have to update the files for this to trigger rebuilding the parser!

## Expected behaviour

It should actually run the scripts and not report it's up to date.

One fix would be to modify the Makefile to always run the script when `parser` and `parser-o` are invoked but I think the problem should instead be found.

## Relevant Output

As I included it in the description ...

## bug report attachment

The relevant parts are running the scripts directly but which do not help the problem since the problem is they're not running normally. I mean maybe but you'll find a lot of compilation errors and missing tools because `jparse` cannot be compiled for some reason.

Anyway the scripts being run do work okay from bug_report.sh but not when running `make` or `make parser-o` or `make parser` which is a big problem:

```sh
$ make run_flex-v7  
./run_flex.sh -v 7 -s ./sorry.tm.ca.h -g ./verge -D .
./run_flex.sh: debug[1]: pass no args to flex before ./jparse.l
./run_flex.sh: debug[5]: FLEX_BASENAME=flex
./run_flex.sh: debug[5]: MIN_FLEX_VERSION=2.6.4
./run_flex.sh: debug[5]: PREFIX=./jparse
./run_flex.sh: debug[5]: RUN_FLEX_VERSION=1.0 2023-02-04
./run_flex.sh: debug[5]: SORRY_H=./sorry.tm.ca.h
./run_flex.sh: debug[5]: VERGE=./verge
./run_flex.sh: debug[5]: V_FLAG=7
./run_flex.sh: debug[3]: checking for flex as: flex
./run_flex.sh: debug[5]: on_path "flex" "2.6.4"
./run_flex.sh: debug[7]: tool: /opt/local/bin/flex has version: 2.6.4
./run_flex.sh: debug[7]: tool: /opt/local/bin/flex version: 2.6.4 >= min version: 2.6.4
./run_flex.sh: debug[3]: tool on $PATH: /opt/local/bin/flex version: 2.6.4 is good
./run_flex.sh: debug[1]: flex on $PATH is good: /opt/local/bin/flex
/opt/local/bin/flex  ./jparse.l
# prepending comment and line number reset to ./jparse.c
cat ./sorry.tm.ca.h > .sorry.run_flex.sh.XXXXXXXXXX.c
cat ./jparse.c >> .sorry.run_flex.sh.XXXXXXXXXX.c
mv -v -f .sorry.run_flex.sh.XXXXXXXXXX.c ./jparse.c
.sorry.run_flex.sh.XXXXXXXXXX.c -> ./jparse.c
# completed update of ./jparse.c
# prepending comment and line number reset to ./jparse.lex.h
cat ./sorry.tm.ca.h > .sorry.run_flex.sh.XXXXXXXXXX.c
cat ./jparse.lex.h >> .sorry.run_flex.sh.XXXXXXXXXX.c
mv -v -f .sorry.run_flex.sh.XXXXXXXXXX.c ./jparse.lex.h
.sorry.run_flex.sh.XXXXXXXXXX.c -> ./jparse.lex.h
# completed update of ./jparse.lex.h
./run_flex.sh: debug[1]: -rw-------  1 cody  staff  95841 Mar  9 08:08 ./jparse.c
./run_flex.sh: debug[1]: -rw-------  1 cody  staff  19400 Mar  9 08:08 ./jparse.lex.h
./run_flex.sh: debug[1]: flex run OK: formed ./jparse.c and ./jparse.lex.h
$ make run_bison-v7
./run_bison.sh -v 7 -s ./sorry.tm.ca.h -g ./verge -D .
./run_bison.sh: debug[1]: pass no args to bison before ./jparse.y
./run_bison.sh: debug[5]: BISON_BASENAME=bison
./run_bison.sh: debug[5]: MIN_BISON_VERSION=3.8.2
./run_bison.sh: debug[5]: PREFIX=./jparse
./run_bison.sh: debug[5]: RUN_BISON_VERSION=1.0 2023-02-04
./run_bison.sh: debug[5]: SORRY_H=./sorry.tm.ca.h
./run_bison.sh: debug[5]: VERGE=./verge
./run_bison.sh: debug[5]: V_FLAG=7
./run_bison.sh: debug[3]: checking for bison as: bison
./run_bison.sh: debug[5]: on_path "bison" "3.8.2"
./run_bison.sh: debug[7]: tool: /opt/local/bin/bison has version: 3.8.2
./run_bison.sh: debug[7]: tool: /opt/local/bin/bison version: 3.8.2 >= min version: 3.8.2
./run_bison.sh: debug[3]: tool on $PATH: /opt/local/bin/bison version: 3.8.2 is good
./run_bison.sh: debug[1]: bison on $PATH is good: /opt/local/bin/bison
/opt/local/bin/bison  ./jparse.y
# prepending comment and line number reset to ./jparse.tab.c
cat ./sorry.tm.ca.h > .sorry.run_bison.sh.XXXXXXXXXX.c
cat ./jparse.tab.c >> .sorry.run_bison.sh.XXXXXXXXXX.c
mv -v -f .sorry.run_bison.sh.XXXXXXXXXX.c ./jparse.tab.c
.sorry.run_bison.sh.XXXXXXXXXX.c -> ./jparse.tab.c
# completed update of ./jparse.tab.c
# prepending comment and line number reset to ./jparse.tab.h
cat ./sorry.tm.ca.h > .sorry.run_bison.sh.XXXXXXXXXX.c
cat ./jparse.tab.h >> .sorry.run_bison.sh.XXXXXXXXXX.c
mv -v -f .sorry.run_bison.sh.XXXXXXXXXX.c ./jparse.tab.h
.sorry.run_bison.sh.XXXXXXXXXX.c -> ./jparse.tab.h
# completed update of ./jparse.tab.h
./run_bison.sh: debug[1]: -rw-------  1 cody  staff  90910 Mar  9 08:08 ./jparse.tab.c
./run_bison.sh: debug[1]: -rw-------  1 cody  staff  13671 Mar  9 08:08 ./jparse.tab.h
./run_bison.sh: debug[1]: bison run OK: formed ./jparse.tab.c and ./jparse.tab.h

```

but these scripts should run when `make parser` is run anyway and iirc we wanted it to run even if nothing has changed. Pretty sure that's how it used to be.

```
NOTE: Being both both **clear** and **concise** helps us.
```

Come now ... do you really expect me to be concise ? :-) .. no but I do think it is good enough.

**Comment by xexyl** [comment]
Oh I should point out that it's not the fact the verbosity is not high enough as I also updated the Makefile: it still doesn't run.

**Comment by xexyl** [comment]
I love how I report the bug and almost immediately you assign it to ME!

I'm honoured but I do think it's very funny.

**Comment by xexyl** [comment]
As for this problem it gets stranger.

Changing the rule in the Makefile to not have any dependencies:

```makefile
jparse.tab.c jparse.tab.h: #jparse.y jparse.l jparse.h sorry.tm.ca.h run_bison.sh verge jparse.tab.ref.c jparse.tab.ref.h

```

and it STILL shows this message!

Unless I'm totally out of my reckoning something is very wrong here. I'm going to try and duplicate it on another system. Maybe a recent change on the macOS system did it .. maybe an installation of make as a dependency of another third party program.

**Comment by xexyl** [comment]
Okay so the script seems to be running though. So why does it not show up? Is it ending before it's supposed to? Putting in some debug echoes.

**Comment by xexyl** [comment]
I know why the verbosity is not working.

Check this:

```
jparse.tab.c jparse.tab.h: jparse.y jparse.h sorry.tm.ca.h run_bison.sh verge jparse.tab.ref.c jparse.tab.ref.h
        ${Q} ./run_bison.sh -v 9 -g ./verge -s sorry.tm.ca.h -b ${BISON_BASENAME} ${BISON_DIRS} \
            -p jparse -v ${Q_V_OPTION} ${RUN_O_FLAG} -- ${BISON_FLAGS}

```

but then `Q_V_OPTION` is not even defined. I put the `-v 9` ahead of it though so it of course was reset to 0. Let's see what happens if I add it after that.

**Comment by xexyl** [comment]
```sh
# completed update of ./jparse.tab.h
./run_bison.sh: debug[1]: -rw------- 1 xexyl xexyl 86798 Mar  9 09:07 ./jparse.tab.c
./run_bison.sh: debug[1]: -rw------- 1 xexyl xexyl 7773 Mar  9 09:07 ./jparse.tab.h
./run_bison.sh: debug[1]: bison run OK: formed ./jparse.tab.c and ./jparse.tab.h
make[1]: 'jparse.tab.h' is up to date.
make[1]: Leaving directory '/home/xexyl/code/xexyl/mkiocccentry/jparse'
make jparse.tab.o
```

... strange. It forms the file and yet it doesn't compile ? Well I'm making progress at least I guess.

**Comment by xexyl** [comment]
Figured it out!

... will update this in a moment ..

**Comment by xexyl** [comment]
In an unpublished commit 18cd69d63aa57b82ce6dfd78094b68f40ad21e0c (HEAD -> jparse):

    Fix up to date issues in make parser rule
    
    An annoying problem was in jparse/Makefile when running make parser that
    made it appear that the scripts were not running. There are two parts to
    this.
    
    First the script was set to 0 verbosity level and I did not realise this
    so it seemed as it was not running (only obscure hints that it was). The
    second part is that the make parser rule had the lines:
    
            ${E} ${MAKE} jparse.tab.c jparse.tab.h
            ${E} ${MAKE} jparse.c jparse.lex.h
    
    and the problem with that is that the first of each implies the second
    so we would get the messages:
    
        make[2]: `jparse.tab.h' is up to date.
    
        make[2]: `jparse.lex.h' is up to date.
    
    This made me believe that the problem I was facing was the scripts were
    not running but instead it's that the rule was run twice which is
    unnecessary. Thus a comment above each one notes that the rule implies
    the other and no longer runs both.

Of course it does not solve the problem I was having as it was actually running. I do have to verify that `make parser-o` is still working correctly but other than that this issue will be said to be resolved once the commit is merged. I'm unsure if I will do that now or later but I'm tired of the other issue for now so I might hold off on that and just make a pull request.

**Comment by xexyl** [comment]
... well I was right about something too: it is not copying the ref file to the other file or so it seems. The ref file has what should be correct that I would think allow compilation but the other file does not. Strange ...

Time to enable verbosity.

**Comment by xexyl** [comment]
For some reason the run_flex.sh is updating the backup file but not the normal one. By updating I mean it doesn't seem to generate the file that's being generated as a backup file. I'm stumped at this time so I'm going to do other things. Shame because once this is resolved I think the prefix is good to go. I might or might not open a new issue - I suspect not though.

Good day!

**Comment by xexyl** [comment]
I've noticed another odd problem. The script is adding the apology file to the `jparse.c` file twice. I'm not sure if it's the fix above or not. I'll see if I can find out.

**Comment by xexyl** [comment]
Closing this as complete as the [commit](https://github.com/ioccc-src/mkiocccentry/pull/513/commits/18cd69d63aa57b82ce6dfd78094b68f40ad21e0c) (https://github.com/ioccc-src/mkiocccentry/pull/513/commits/18cd69d63aa57b82ce6dfd78094b68f40ad21e0c) was merged (thanks!).
